### PR TITLE
feat: add Pydantic AI integration with Composio.

### PR DESCRIPTION
**Pull request by alt-glitch**
> [!IMPORTANT]
> Adds Pydantic AI integration with Composio for type-safe GitHub operations, including setup, demo, documentation, and testing.
> 
>   - **Integration**:
>     - Introduces Pydantic AI integration with Composio for automating GitHub operations using type-safe tools.
>     - Provides a `ComposioToolSet` class in `toolset.py` for wrapping Composio tools as Pydantic-AI Tool objects.
>   - **Setup and Demo**:
>     - Adds `setup.py` for package configuration.
>     - Includes `pydantic_ai_demo.py` demonstrating the integration with a task to star a GitHub repository.
>   - **Documentation**:
>     - Adds `README.md` with installation, setup, and usage instructions for the integration.
>   - **Testing**:
>     - Updates `test_example.py` to include a test case for the Pydantic AI integration demo.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ComposioHQ%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for d46fce7ce2727584bf8cd35ada2f9eaf9d584590. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/plugins/pydanticai/pydantic_ai_demo.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+composio_toolset = ComposioToolSet()
+
+# Get GitHub tools that are pre-configured
+tools = composio_toolset.get_tools(
+    actions=[Action.GITHUB_STAR_A_REPOSITORY_FOR_THE_AUTHENTICATED_USER]
+)
+
+# Create an agent with the tools
+agent = Agent(
+    model="openai:gpt-4o",  # Using a known model name
```
Comment (shreysingla11) [inline python/plugins/pydanticai/pydantic_ai_demo.py:25]: The model name 'openai:gpt-4o' appears to be incorrect. It should be 'openai:gpt-4-turbo' to match the example in README.md. Please update for consistency.

File: python/plugins/pydanticai/pydantic_ai_demo.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    You can perform various GitHub operations using the available tools.
+    When given a task, analyze it and use the appropriate tool to complete it.""",
+)
+
+
+def main():
+    # Define task
+    task = "Star a repo wjayesh/mahilo on GitHub"
+
+    # Run the agent
```
Comment (shreysingla11) [inline python/plugins/pydanticai/pydantic_ai_demo.py:37]: Consider adding error handling around the agent.run_sync() call to gracefully handle potential API errors or rate limits. Example:
```python
try:
    result = agent.run_sync(task)
    print("Result: ", result.data)
    print("Trace:\n\n", result.all_messages())
except Exception as e:
    print(f"Error executing task: {e}")
```

File: python/plugins/pydanticai/pydantic_ai_demo.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+agent = Agent(
+    model="openai:gpt-4o",  # Using a known model name
+    tools=tools,
+    system_prompt="""You are an AI agent that helps users interact with GitHub.
+    You can perform various GitHub operations using the available tools.
+    When given a task, analyze it and use the appropriate tool to complete it.""",
+)
+
+
+def main():
```
Comment (shreysingla11) [inline python/plugins/pydanticai/pydantic_ai_demo.py:33]: Consider adding a docstring for the main() function to explain its purpose and functionality. Example:
```python
def main():
    """
    Main function that demonstrates the usage of Pydantic AI agent with Composio.
    
    Executes a sample task to star a GitHub repository using the configured agent.
    Prints the execution result and trace information.
    """
```

File: python/plugins/pydanticai/composio_pydanticai/toolset.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    )
+
+    # Define task
+    task = "Star a repo wjayesh/mahilo on GitHub"
+    # Run the agent
+    result = agent.run_sync(task)
+    ```
+    """
+
+    def _wrap_action(
```
Comment (shreysingla11) [inline python/plugins/pydanticai/composio_pydanticai/toolset.py:64]: Consider adding type hints for the schema_params dictionary parameter to improve code clarity:
```python
def _wrap_action(
    self,
    action: str,
    description: str,
    schema_params: t.Dict[str, t.Any],
    entity_id: t.Optional[str] = None,
) -> t.Callable:
```

File: python/plugins/pydanticai/composio_pydanticai/toolset.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            function=action_func,
+            name=action,
+            description=description,
+            takes_ctx=True,
+            max_retries=3,  # Can be made configurable
+            docstring_format="auto",
+        )
+    
+
+    def get_tools(
```
Comment (shreysingla11) [inline python/plugins/pydanticai/composio_pydanticai/toolset.py:168]: Consider adding input validation for the actions parameter to ensure all provided actions are valid before processing. This would help catch configuration errors early.

File: python/tests/test_example.py [hunk 6]
```diff
@@ -186,6 +186,15 @@
     #     "env": {"COMPOSIO_API_KEY": COMPOSIO_API_KEY},
     # },
     # TODO: Add camel
+    "pydantic_ai": {
```
Comment (shreysingla11) [inline python/tests/test_example.py:189]: Consider adding more test cases to cover error scenarios and edge cases. Current test only verifies successful execution. Some suggestions:
- Test with invalid model names
- Test with invalid actions
- Test error handling
- Test rate limiting scenarios

**Comment by shreysingla11** [comment]
I've reviewed the Pydantic AI integration PR and here are my main observations:

1. **Code Quality**:
   - Overall good structure and organization of the plugin
   - Well-documented README with clear examples
   - Good separation of concerns between different components

2. **Suggested Improvements**:
   - Fix inconsistent model names between demo and README
   - Add more comprehensive error handling
   - Enhance documentation with additional docstrings
   - Add more test coverage for error scenarios
   - Consider adding input validation for critical parameters

3. **Security & Best Practices**:
   - Good use of environment variables for sensitive data
   - Consider adding rate limiting handling
   - Add input sanitization for user-provided values

The implementation is solid and well-structured. I've added specific comments to relevant sections of the code with suggested improvements.

Please review the inline comments for detailed suggestions.

File: python/plugins/pydanticai/setup.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+from pathlib import Path
+
+from setuptools import setup
+
+
+setup(
+    name="composio_pydanticai",
+    version="0.6.12",
+    author="Siddharth",
+    author_email="sid@composio.dev",
```
Comment (angrybayblade) [inline python/plugins/pydanticai/setup.py:14]: Use `tech@composio.dev` account instead

File: python/plugins/pydanticai/README.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+from composio_pydanticai import ComposioToolSet
+from pydantic_ai import Agent
+```
+
+#### 2. Initialize Tools with Composio
+
+Configure and fetch GitHub tools provided by Composio:
+
+```python
+from composio_pydanticai import ComposioToolSet, ToolConfig
```
Comment (angrybayblade) [inline python/plugins/pydanticai/README.md:45]: `ToolConfig` is not a top level export, fix this

File: python/plugins/pydanticai/README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+# Initialize tools with default retry settings
+composio_toolset = ComposioToolSet(max_retries=3)  # Default max retries for all tools
+
+# Or configure specific tools differently
+tool_configs = {
+    "github_star_repository": ToolConfig(max_retries=5),  # More retries for starring
+    "github_create_repository": ToolConfig(max_retries=2)  # Fewer retries for creation
+}
+composio_toolset = ComposioToolSet(max_retries=3, tool_configs=tool_configs)
```
Comment (angrybayblade) [inline python/plugins/pydanticai/README.md:55]: This is anti-pattern for toolset class, you're not supposed to define any top-level configs. Tool configs should be provided when calling `get_tools`

File: python/plugins/pydanticai/composio_pydanticai/toolset.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+@dataclass
+class ToolConfig:
+    """Configuration for a specific tool."""
+
+    max_retries: int = 3
+
+
+class ComposioToolSet(
+    BaseComposioToolSet,
+    t.Generic[AgentDeps],
```
Comment (angrybayblade) [inline python/plugins/pydanticai/composio_pydanticai/toolset.py:31]: What's the point of using `Generic` here?

File: python/plugins/pydanticai/composio_pydanticai/toolset.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        self,
+        action: str,
+        description: str,
+        schema_params: t.Dict[str, t.Any],
+        entity_id: t.Optional[str] = None,
+    ):
+        """Create a wrapper function for the Composio action."""
+
+        async def function(
+            ctx: "RunContext[AgentDeps]", **kwargs: t.Any
```
Comment (angrybayblade) [inline python/plugins/pydanticai/composio_pydanticai/toolset.py:102]: This does not mean anything unless the type annotations are delegated outside of the wrapper function

**Comment by alt-glitch** [comment]
closes #1193

File: python/plugins/pydanticai/setup.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+"""
+
+from pathlib import Path
+
+from setuptools import setup
+
+
+setup(
+    name="composio_pydanticai",
+    version="0.6.12",
```
Comment (angrybayblade) [inline python/plugins/pydanticai/setup.py:12]:
```suggestion
    version="0.6.16",
```

**Review by angrybayblade — APPROVED** [review]
